### PR TITLE
rbd: return error if last sync time not present

### PR DESCRIPTION
**Pull request by Madhu-1**
As per the csiaddon spec, last sync time is a required parameter in the GetVolumeReplicationInfo if we are failed to parse the description, we return an error message instead of nil which is an empty response.

Signed-off-by: Madhu Rajanna [madhupr007@gmail.com](mailto:madhupr007@gmail.com)

Note:- skipping E2E as we dont have CI for DR

Note:- Replacement for https://github.com/ceph/ceph-csi/pull/3478 after discussion in csiaddons spec.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -27,7 +27,7 @@ require (
 	github.com/pkg/xattr v0.4.7
 	github.com/prometheus/client_golang v1.12.2
 	github.com/stretchr/testify v1.8.1
-	golang.org/x/crypto v0.0.0-20220315160706-3147a52a75dd
+	golang.org/x/crypto v0.1.0
```
Comment (Madhu-1) [inline go.mod]: this got updated as part of `go get`

**Review by ShyamsundarR — COMMENTED** [review]
LGTM

**Comment by Madhu-1** [comment]
@ShyamsundarR PTAL at last commit. This check was missing, and cephcsi was returning an empty response.

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by ShyamsundarR** [comment]
> @ShyamsundarR PTAL at last commit. This check was missing, and cephcsi was returning an empty response.

Thanks, I am only looking at overall structural correctness (hence not hitting the approved button). @yati1998 assuming you are taking a closer look?

File: internal/rbd/replicationcontrollerserver.go [hunk 2]
```diff
@@ -804,13 +809,14 @@ func getLastSyncTime(description string) (*timestamppb.Timestamp, error) {
 	// description = "replaying,{"bytes_per_second":0.0,
 	// "bytes_per_snapshot":149504.0,"local_snapshot_timestamp":1662655501
 	// ,"remote_snapshot_timestamp":1662655501}"
-	// In case there is no local snapshot timestamp we can pass the default value
+	// In case there is no local snapshot timestamp return an error as the
+	// LastSyncTime is required.
 	if description == "" {
-		return nil, nil
+		return nil, fmt.Errorf("%w empty description", ErrLastSyncTimeNotFound)
```
Comment (nixpanic) [inline internal/rbd/replicationcontrollerserver.go]: missing `:` after `%w`?

File: internal/rbd/replicationcontrollerserver.go [hunk 3]
```diff
@@ -804,13 +809,14 @@ func getLastSyncTime(description string) (*timestamppb.Timestamp, error) {
...
 	}
 	splittedString := strings.SplitN(description, ",", 2)
 	if len(splittedString) == 1 {
-		return nil, nil
+		return nil, fmt.Errorf("%w no local snapshot timestamp", ErrLastSyncTimeNotFound)
```
Comment (nixpanic) [inline internal/rbd/replicationcontrollerserver.go]: same here, formatting of error messages matter

**Review by nixpanic — CHANGES_REQUESTED** [review]
I think this looks good, but (a nit) the formatting of the error messages should be nicer.

**Review by nixpanic — DISMISSED** [review]

**Review by yati1998 — APPROVED** [review]
LGTM

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Review by Rakshith-R — APPROVED** [review]

**Review by nixpanic — APPROVED** [review]

**Pull request merged**